internal/mode/static/telemetry/collector_test.go

```diff
@@ -0,0 +1,338 @@
+
+		dataCollector = telemetry.NewDataCollectorImpl(telemetry.DataCollectorConfig{
+			K8sClientReader:     k8sClientReader,
+			GraphGetter:         fakeGraphGetter,
+			ConfigurationGetter: fakeConfigurationGetter,
+			Version:             version,
+		})
+	})
+
+	Describe("Normal case", func() {
+		When("collecting telemetry data", func() {
+			It("collects all fields", func() {
+				nodes := []v1.Node{
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "node1"},
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "node2"},
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "node3"},
+					},
+				}
+
+				k8sClientReader.ListCalls(createListCallsFunc(nodes))
+
+				secret1 := &v1.Secret{ObjectMeta: metav1.ObjectMeta{Name: "secret1"}}
+				secret2 := &v1.Secret{ObjectMeta: metav1.ObjectMeta{Name: "secret2"}}
+				nilsecret := &v1.Secret{ObjectMeta: metav1.ObjectMeta{Name: "nilsecret"}}
+
+				svc1 := &v1.Service{ObjectMeta: metav1.ObjectMeta{Name: "svc1"}}
+				svc2 := &v1.Service{ObjectMeta: metav1.ObjectMeta{Name: "svc2"}}
+				nilsvc := &v1.Service{ObjectMeta: metav1.ObjectMeta{Name: "nilsvc"}}
+
+				graph := &graph.Graph{
+					GatewayClass: &graph.GatewayClass{},
+					Gateway:      &graph.Gateway{},
+					IgnoredGatewayClasses: map[types.NamespacedName]*gatewayv1.GatewayClass{
+						{Name: "ignoredGC1"}: {},
+						{Name: "ignoredGC2"}: {},
+					},
+					IgnoredGateways: map[types.NamespacedName]*gatewayv1.Gateway{
+						{Name: "ignoredGw1"}: {},
+						{Name: "ignoredGw2"}: {},
+					},
+					Routes: map[types.NamespacedName]*graph.Route{
+						{Namespace: "test", Name: "hr-1"}: {},
+						{Namespace: "test", Name: "hr-2"}: {},
+						{Namespace: "test", Name: "hr-3"}: {},
+					},
+					ReferencedSecrets: map[types.NamespacedName]*graph.Secret{
+						client.ObjectKeyFromObject(secret1): {
+							Source: secret1,
+						},
+						client.ObjectKeyFromObject(secret2): {
+							Source: secret2,
+						},
+						client.ObjectKeyFromObject(nilsecret): nil,
+					},
+					ReferencedServices: map[types.NamespacedName]struct{}{
+						client.ObjectKeyFromObject(svc1):   {},
+						client.ObjectKeyFromObject(svc2):   {},
+						client.ObjectKeyFromObject(nilsvc): {},
+					},
+				}
+
+				config := &dataplane.Configuration{
+					Upstreams: []dataplane.Upstream{
+						{
+							Name:     "upstream1",
+							ErrorMsg: "",
+							Endpoints: []resolver.Endpoint{
+								{
+									Address: "endpoint1",
+									Port:    80,
+								}, {
+									Address: "endpoint2",
+									Port:    80,
+								}, {
+									Address: "endpoint3",
+									Port:    80,
+								},
+							},
+						},
+						{
+							Name:     "upstream2",
+							ErrorMsg: "",
+							Endpoints: []resolver.Endpoint{
+								{
+									Address: "endpoint1",
+									Port:    80,
+								},
+							},
+						},
+						{
+							Name:     "upstream3",
+							ErrorMsg: "there is an error here",
```

Comment:
if this is an edge case, I'd recommend moving it into a separate edge case test.